### PR TITLE
Port flat PA from habana_next to habana_main

### PR DESCRIPTION
**Pull request by dolszewska**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/worker/habana_model_runner.py [hunk 1]
```diff
@@ -147,45 +151,62 @@ def align_workers(value, op):
...
+        attn_bias = (torch.zeros_like(mask, dtype=dtype)
+                        .masked_fill_(mask, -math.inf))
+        return metadata._replace(attn_bias=attn_bias)
+
+    def _set_block_mapping(self, metadata, batch_size, device, dtype):
+        mask = torch.arange(0, self.block_size, device=device, dtype=torch.int32).unsqueeze(0)
+        mask = mask >= metadata.block_usage.unsqueeze(-1)
+        attn_bias = (torch.zeros_like(mask, dtype=dtype)
+                        .masked_fill_(mask, -math.inf))
+        block_mapping = torch.nn.functional.one_hot(metadata.block_mapping, num_classes=batch_size).to(dtype)
```
Comment (libinta) [inline vllm/worker/habana_model_runner.py:190]: the input to one_hot should be cast to torch.long, otherwise, it has runtime error

File: vllm/worker/habana_model_runner.py [hunk 2]
```diff
@@ -785,15 +786,15 @@ def _prepare_decode(
             assert seq_group_metadata.token_chunk_size == 1
 
             seq_ids = list(seq_group_metadata.seq_data.keys())
-            lora_id = seq_group_metadata.lora_int_id
+            #lora_id = seq_group_metadata.lora_int_id
 
-            if lora_id > 0:
-                lora_requests.add(seq_group_metadata.lora_request)
+            #if lora_id > 0:
+            #    lora_requests.add(seq_group_metadata.lora_request)
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:792]: we have LoRA support now and this will break it

File: vllm/worker/habana_model_runner.py [hunk 3]
```diff
@@ -808,8 +809,8 @@ def _prepare_decode(
                 block_offset = position % self.block_size
                 slot = block_number * self.block_size + block_offset
                 slot_mapping.append([slot])
-                lora_index_mapping.append(lora_id)
-                lora_prompt_mapping.append(lora_id)
+                #lora_index_mapping.append(lora_id)
+                #lora_prompt_mapping.append(lora_id)
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:813]: we have LoRA support now and this will break it

**Review by kzawora-intel — CHANGES_REQUESTED** [review]

File: vllm/worker/habana_model_runner.py [hunk 4]
```diff
@@ -493,64 +536,47 @@ def _is_valid_bucket(self, bucket):
...
-                                                          max=2048)
+                                                          max=max_prompt_seq)
+        self.decode_block_bucket_cfg = read_bucket_settings(
+            'decode',
+            'block',
+            min=blocks_step,
+            step=blocks_step,
+            max=max(blocks_step,
+                    self.max_num_seqs * max_decode_seq // self.block_size))
         self.graphed_buckets: Set[Any] = set()
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:570]: this changes ``VLLM_DECODE_SEQ_BUCKET_{p}`` env vars into ``VLLM_DECODE_BLOCK_BUCKET_{p}`` - such a change needs to be documented in gaudi READMEs - the user is not aware of that and could still use previous knobs to no effect

File: vllm/worker/habana_model_runner.py [hunk 5]
```diff
@@ -90,18 +108,31 @@ def warmup_range(config: Tuple[int, int, int]):
...
+                            max_blocks):
+    buckets = []
+    for bs in warmup_range(bs_bucket_config):
+        for blocks in warmup_range(blocks_bucket_config):
+            if blocks < bs:
+                continue
+            if blocks > max_blocks:
+                break
+            buckets.append((bs, blocks))
+    return list(sorted(buckets, key=lambda b: (b[0] * b[1], b[1], b[0])))
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:131]: Token budget was introduced in https://github.com/HabanaAI/vllm-fork/pull/206 and these functions should take the budget into account (filter out "too expensive" buckets)

File: vllm/worker/habana_model_runner.py [hunk 6]
```diff
@@ -493,64 +537,47 @@ def _is_valid_bucket(self, bucket):
...
-        self.prompt_bs_bucket_cfg = read_bucket_settings('prompt',
-                                                         'bs',
-                                                         min=1,
-                                                         step=32,
-                                                         max=min(
-                                                             self.max_num_seqs,
-                                                             max_bucket_cfg))
+        blocks_step = 128
+        max_prompt_seq = 1024
+        max_decode_seq = 2048
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:547]: let's not have these hardcoded values as defaults, and use max_model_len instead

File: vllm/worker/habana_model_runner.py
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:1]: caution: `prepare_input_tensors` is affected by https://github.com/HabanaAI/vllm-fork/issues/158; you need to modify the following code not to duplicate first sequence's block tables, as that will give you incorrect block sums:
```python
        seq_group_metadata_list = seq_group_metadata_list.copy()
        seq_group_metadata_list.extend(seq_group_metadata_list[0]
                                       for _ in range(batch_size_padding))
```
please sync with @iboiko-habana who prepares a fix for that

File: vllm/worker/habana_model_runner.py [hunk 7]
```diff
@@ -493,64 +537,47 @@ def _is_valid_bucket(self, bucket):
...
-        self.decode_buckets = warmup_buckets(self.decode_bs_bucket_cfg,
-                                             self.decode_seq_bucket_cfg)
-        if self.lora_config:
-            self.decode_buckets[:] = [
-                bucket for bucket in self.decode_buckets
-                if self._is_valid_bucket(bucket)
-            ]
-        msg = (f"Generated {len(self.decode_buckets)} decode buckets: "
-               f"{list(sorted(self.decode_buckets))}")
+               f"seq:{self.decode_block_bucket_cfg}")
```
Comment (kzawora-intel) [inline vllm/worker/habana_model_runner.py:580]: flat pa does not use seq bucket config
```suggestion
               f"block:{self.decode_block_bucket_cfg}")
```

File: vllm/worker/habana_model_runner.py
Comment (dolszewska) [reply vllm/worker/habana_model_runner.py:1]: I saw that Iryna's fix is ready for habana_next and I think it can be cherry-pick after flat-PA is merged

File: vllm/hpu/ops.py [hunk 8]
```diff
@@ -29,72 +28,57 @@
...
+    return (block_mapping @ tensor.view(shape[0], -1)).view(-1, *shape[1:])
+
+
+def block2batch(tensor, block_mapping):
+    shape = tuple(tensor.shape)
+    return (block_mapping.t() @ tensor.view(shape[0], -1)).view(-1, *shape[1:])
+
+
+def block_softmax(batch_size, attn, block_mapping):
+    attn.sub_(10.0)
```
[the remark this answers is not in the document]
Comment (dolszewska) [reply vllm/hpu/ops.py:43]: The fix was tested and it does not help for accuracy in Llama

File: vllm/worker/habana_model_runner.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dolszewska) [reply vllm/worker/habana_model_runner.py:179]: max_num_batched_tokens was removed for decode buckets

**Review by madamczykhabana — APPROVED** [review]
LGTM
